pkg/util/subnet/subnet.go

```diff
@@ -84,12 +84,17 @@ func NetworkSecurityGroupID(oc *api.OpenShiftCluster, subnetID string) (string,
 		infraID = "aro"
 	}
 
-	switch {
-	case strings.EqualFold(subnetID, oc.Properties.MasterProfile.SubnetID):
+	if strings.EqualFold(subnetID, oc.Properties.MasterProfile.SubnetID) {
 		return oc.Properties.ClusterProfile.ResourceGroupID + "/providers/Microsoft.Network/networkSecurityGroups/" + infraID + NSGControlPlaneSuffix, nil
-	case strings.EqualFold(subnetID, oc.Properties.WorkerProfiles[0].SubnetID):
-		return oc.Properties.ClusterProfile.ResourceGroupID + "/providers/Microsoft.Network/networkSecurityGroups/" + infraID + NSGNodeSuffix, nil
-	default:
-		return "", fmt.Errorf("unknown subnetID %q", subnetID)
 	}
+
+	// node pools may have different subnets, but they share the same NSG
```

Comment:
I might have made up this rule to make things easier. The rational behind it is that customer cannot create new NSGs in the cluster resource group due to its locked down nature. I haven't seen anything inside OpenShift platform that may create a new NSG. 
OpenShift may update existing NSG for example when creating an LB service `oc expose deploy ... --type LoadBalancer`

Reply:
The Azure cloud provider expects all subnets to share an NSG: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L136

Add the above text to the code comment please.

However AFAICS OpenShift 4.4 -> 4.5 changes its architecture: clusters installed by 4.4 have 2 NSGs; only 1 in 4.5.  So either you or @mjudeikis are going to have something to fix up, depending on merge order.

Reply:
I will remove the comment. The config struct also has only one `SubnetName` https://github.com/kubernetes/kubernetes/blob/87cc9bca0fcbbdf1b8232c9c71e50023d62667c3/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L134 but we currently have 2 - master/worker.

Reply:
However I think we should look at the validation not from k8s design perspective, but from perspective of what is possible to do with the cluster and infrastructure, including (and especially) things that render cluster setup "unsupported" or wrong. I have tested that you can create a few worker pools each on its own subnet and the setup works perfectly fine. Maybe the better approach would be to add this validation before the the subnet/routetable validation as you suggested in the earlier comment, and after that we will continue with validation knowing as in original code before these proposed changes.

Reply:
> we currently have 2 - master/worker

In 4.4, yes - workload on the masters is not reachable by a Service of type LoadBalancer.

This is "fixed" in 4.5 clusters, which only have 1 NSG for masters and workers AFAIK.

Reply:
> However I think we should look at the validation not from k8s design perspective, but from perspective of what is possible to do with the cluster and infrastructure, including (and especially) things that render cluster setup "unsupported" or wrong. I have tested that you can create a few worker pools each on its own subnet and the setup works perfectly fine. Maybe the better approach would be to add this validation before the the subnet/routetable validation as you suggested in the earlier comment, and after that we will continue with validation knowing as in original code before these proposed changes.

I don't understand this comment, please could you clarify?

Reply:
You are saying that azure-provider expects all subnets to share NSG, but from the same linked struct I can see that it seems like it is also expecting to have only 1 subnet, which is not true. The cluster can have many subnets, even if cloud provider is not expecting it, but we should check for this scenario. 
The cloud provider that you linked is legacy, is it what we are using in ARO? Currently azure cloud provider is developed in its own repo: https://github.com/kubernetes-sigs/cloud-provider-azure